internal/storage/storage_image.go

```diff
@@ -636,9 +701,177 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
 	return nil, errors.New("blob not found")
 }
 
-func (s *storageImageDestination) Commit(ctx context.Context, unparsedToplevel types.UnparsedImage) error {
+// queueOrCommit queues in the specified blob to be committed to the storage.
+// If no other goroutine is already committing layers, the layer and all
+// subsequent layers (if already queued) will be committed to the storage.
+func (s *StorageImageDestination) queueOrCommit(ctx context.Context, blob types.BlobInfo, index int) error {
+	// NOTE: whenever the code below is touched, make sure that all code
+	// paths unlock the lock and to unlock it exactly once.
+	//
+	// Conceptually, the code is divided in two stages:
+	//
+	// 1) Queue in work by marking the layer as ready to be committed.
+	//    If at least one previous/parent layer with a lower index has
+	//    not yet been committed, return early.
+	//
+	// 2) Process the queued-in work by committing the "ready" layers
+	//    in sequence.  Make sure that more items can be queued-in
+	//    during the comparatively I/O expensive task of committing a
+	//    layer.
```

Comment:
I considered the previous design where each caller would block until it's parent layers have been committed and then commit the layer itself.  I *think* Docker does something similar since Docker's progress bars are showing "downloaded" and "committed" message (not sure about the exact wording but conceptually they do that).

What I like about this design is:
* We have at most 6 goroutines pulling down layers in parallel
* Assume we have > 6 layers to pull

With the old design, 5 of the 6 goroutines may be blocked always waiting until they are allowed to commit. With the new design, only 1 goroutine is the "worker" committing the layers.  The other 5 can continue pulling layers and queue in work.

I like the simplicity of this design and found the code comparatively approachable.

Extended the comment to make this aspect explicit.

Reply:
> With the old design, 5 of the 6 goroutines may be blocked always waiting until they are allowed to commit. With the new design, only 1 goroutine is the "worker" committing the layers. The other 5 can continue pulling layers and queue in work.

That’s a great point, thanks for writing that down.